File: solutions/647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:
    def countSubstrings(self, s: str) -> int:
        palindroms = 0

        def expand(left, right):
            count = 0
            while left >= 0 and right < len(s) and s[left] == s[right]:
                count += 1
                left -= 1
                right += 1
            return count

        for index in range(len(s)):
            palindroms += expand(index, index)
            palindroms += expand(index, index + 1)

        return palindroms

# @lc code=end

File: solutions/647.palindromic-substrings.java
/*
 * @lc app=leetcode id=647 lang=java
 *
 * [647] Palindromic Substrings
 */

// @lc code=start
class Solution {
    public int countSubstrings(String s) {
        int palindroms = 0;
        for(int index = 0; index < s.length(); index++) {
            palindroms += expand(s, index, index);
            palindroms += expand(s, index, index + 1);
        }
        return palindroms;
    }

    private int expand(String s, int left, int right) {
        int counter = 0;
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            counter++;
            left--;
            right++;
        }
        return counter;
    }
}
// @lc code=end

